### DISCUSSION TITLE
Temporal Constraint in FLUSS/FLOSS

### DISCUSSION THREAD

**Discussion by LouisCarpentier42** [TDAmeritrade/stumpy#824] [Help: Coding & Implementations]
The paper that introduces FLOSS [1] for time series segmentation mentions the possibility of a temporal constraint in Section 3.5. This is especially useful in time series with reoccurring behavior, because the Matrix Profile can be confused by a similar segment at a large temporal distance. A temporal constraint limits the horizon of possible neighbor, and thus copes with the problem of reoccurring behavior. 

I was wondering if this temporal constraint is implemented in `stumpy`, and if so, where exactly. I have not found any tracks of including such a temporal constraint, while I think this could be extremely useful for many applications. Or is there a simple way for me to automatically calculate this constrained Matrix Profile? Additionally, FLOSS and FLUSS use the corrected arc crossings (CAC), which basically dictates how many 'nearest neighbors of a segment' cross a specific index. When a temporal constraint is incorporated, then the normalization of this CAC must change. Is this implemented in the libarary? 

Thanks in advance!

[1] S. Gharghabi et al. “Domain agnostic online semantic segmentation for multi-dimensional time series”. In: Data mining and nowledge discovery 33 (2019),pp. 96–130. doi: 10.1007/s10618-018-0589-3 [Link](https://link.springer.com/article/10.1007/s10618-018-0589-3)

**Comment by seanlaw** [comment]
@LouisCarpentier42 Thank you for your question and welcome to the STUMPY community!

First, please note that the current implementation of FLOSS/FLUSS only supports 1D time series (we based our implementation on [this published work](https://www.cs.ucr.edu/~eamonn/Segmentation_ICDM.pdf)). Multi-dimensional data tends to be really tricky and so we haven't looked into this as of yet.

> I was wondering if this temporal constraint is implemented in stumpy, and if so, where exactly. I have not found any tracks of including such a temporal constraint, while I think this could be extremely useful for many applications. Or is there a simple way for me to automatically calculate this constrained Matrix Profile?

No, temporal constraint has not been implemented in STUMPY and there are no plans to.

> I was wondering if this temporal constraint is implemented in stumpy, and if so, where exactly. I have not found any tracks of including such a temporal constraint, while I think this could be extremely useful for many applications. Or is there a simple way for me to automatically calculate this constrained Matrix Profile? Additionally, FLOSS and FLUSS use the corrected arc crossings (CAC), which basically dictates how many 'nearest neighbors of a segment' cross a specific index. When a temporal constraint is incorporated, then the normalization of this CAC must change. Is this implemented in the libarary?

It's been a long time since I've looked at FLOSS/FLUSS and I haven't looked at the paper that you referenced in detail but it sounds like their proposal is purely speculative and it will be really hard to replicate and support. In other words, given the lack of information in the paper, the STUMPY community cannot provide much, if any, technical support (it would be the blind leading the blind). 

Having said that, there seems to be a small hint in the paper that you referenced that says:

> Recall that we correct the IAC to the CAC based on the assumption that in a time series with no locality structure, the arcs from each subsequence point to an effectively random location. However, when using the temporal constraint, the arcs cannot point to any arbitrary location. Thus, the previous assumption is no longer useful here. Nonetheless, here the correction “curve”, instead of being parabolic or a beta distribution, is simply a uniform distribution, except for the first TC×samplingrate, and last TC×sampling-rate data points. As these are asymptotically irrelevant, as shown in Fig. 13, we simply hardcode the corresponding CAC to one in these regions.
Note that temporal constraints require you to make some assumptions about the domain in question. This experiment suggests that if your assumptions are reasonable, this algorithm will work well. If your assumptions are strongly violated, we make no claims.

Indeed, if you look at `floss.py`, you'll see that our default `_iac` function:

> Compute the bidirectional idealized arc curve (IAC). This is based on a beta distribution that is scaled with a width that is identical to the length of the matrix profile index. The height of the idealized parabolic curve is assumed to be exactly half the width.

which is consistent with what is mentioned in the paper and basically defines a set of weights to apply to your counts of overhead crossing/arcs. The way that we've handled correcting this arc curve is to allow you, the user, to either apply the default IAC correction OR to specify your own in `stumpy.floss` or `stumpy.fluss` via the `custom_iac` parameter, which is a `numpy` array of weights and the length of this array, I think, should be equal to `len(I)`.  Again, this is already beyond the edge of my understanding/comfort zone but, hopefully, you are able to make more sense of it than I can.
